internal/internal_event_handlers.go

```diff
@@ -970,9 +970,17 @@ func (weh *workflowExecutionEventHandlerImpl) handleActivityTaskTimedOut(event *
 		return nil
 	}
 
+	var err error
 	attributes := event.ActivityTaskTimedOutEventAttributes
-	details := newEncodedValues(attributes.Details, weh.GetDataConverter())
-	err := NewTimeoutError(attributes.GetTimeoutType(), details)
+	if len(attributes.GetLastFailureReason()) > 0 {
+		// When retry activity timeout, it is possible that previous attempts got other customer timeout errors.
+		// To stabilize the error type, we always return the customer error.
+		// See more details of background: https://github.com/uber/cadence/issues/2627
+		err = constructError(attributes.GetLastFailureReason(), attributes.LastFailureDetails, weh.GetDataConverter())
```

Comment:
the last error reason could be timeout as well, could you test that as well?

Reply:
addressed offline. Our test case will cover it.